<a href="https://colab.research.google.com/github/universeatmyfeet/DL/blob/master/Lab3/Keras_Optimizer_Lab3_J051.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Usage of optimizers

An optimizer is one of the two arguments required for compiling a Keras model:

The optimizer is provided in model compilation step as shown below.

In [0]:
import keras
from keras import models
from keras.layers import Dense, Dropout
from keras.utils import to_categorical
from keras.datasets import mnist
from keras.utils.vis_utils import model_to_dot
from IPython.display import SVG

NUM_ROWS = 28
NUM_COLS = 28
NUM_CLASSES = 10
BATCH_SIZE = 128
EPOCHS = 10

# Load data
(X_train, y_train), (X_test, y_test) = mnist.load_data()



# Reshape data
X_train = X_train.reshape((X_train.shape[0], NUM_ROWS * NUM_COLS))
X_train = X_train.astype('float32') / 255
X_test = X_test.reshape((X_test.shape[0], NUM_ROWS * NUM_COLS))
X_test = X_test.astype('float32') / 255

# Categorically encode labels
y_train = to_categorical(y_train, NUM_CLASSES)
y_test = to_categorical(y_test, NUM_CLASSES)


# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='relu', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='relu'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

#Parameters common to all Keras optimizers

The parameters clipnorm and clipvalue can be used with all optimizers to control gradient clipping:

#SGD

keras.optimizers.SGD(learning_rate=0.01, momentum=0.0, nesterov=False)

##Stochastic gradient descent optimizer.

Includes support for momentum, learning rate decay, and Nesterov momentum.

##Arguments

1. learning_rate: float >= 0. Learning rate.
2. momentum: float >= 0. Parameter that accelerates SGD in the relevant direction and dampens oscillations.
3. nesterov: boolean. Whether to apply Nesterov momentum.

In [0]:
from keras import optimizers
sgd=keras.optimizers.SGD(lr=0.01, momentum=0.0, nesterov=False)

In [0]:
# Compile model using above optimizer
model.compile(optimizer=sgd,
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

#RMSprop

keras.optimizers.RMSprop(learning_rate=0.001, rho=0.9)

##RMSProp optimizer.

It is recommended to leave the parameters of this optimizer at their default values (except the learning rate, which can be freely tuned).

Arguments

1. learning_rate: float >= 0. Learning rate.
2. rho: float >= 0.

In [0]:
from keras import optimizers
rms=keras.optimizers.RMSprop(lr=0.001, rho=0.9)

In [0]:
# Compile model using above optimizer
model.compile(optimizer=rms,
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

#Adagrad

keras.optimizers.Adagrad(learning_rate=0.01)

##Adagrad optimizer.

Adagrad is an optimizer with parameter-specific learning rates, which are adapted relative to how frequently a parameter gets updated during training. The more updates a parameter receives, the smaller the learning rate.

It is recommended to leave the parameters of this optimizer at their default values.

##Arguments

1. learning_rate: float >= 0. Initial learning rate.

In [0]:
from keras import optimizers
ada=keras.optimizers.Adagrad(lr=0.01)


In [0]:
# Compile model using above optimizer
model.compile(optimizer=ada,
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

#Adadelta

keras.optimizers.Adadelta(learning_rate=1.0, rho=0.95)

##Adadelta optimizer.

Adadelta is a more robust extension of Adagrad that adapts learning rates based on a moving window of gradient updates, instead of accumulating all past gradients. This way, Adadelta continues learning even when many updates have been done. Compared to Adagrad, in the original version of Adadelta you don't have to set an initial learning rate. In this version, initial learning rate and decay factor can be set, as in most other Keras optimizers.

It is recommended to leave the parameters of this optimizer at their default values.

##Arguments

1. learning_rate: float >= 0. Initial learning rate, defaults to 1. It is recommended to leave it at the default value.
2. rho: float >= 0. Adadelta decay factor, corresponding to fraction of gradient to keep at each time step.

In [0]:
from keras import optimizers
adelta=keras.optimizers.Adadelta(lr=1.0, rho=0.95)



In [0]:
# Compile model using above optimizer
model.compile(optimizer=adelta,
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

#Adam

keras.optimizers.Adam(learning_rate=0.001, beta_1=0.9, beta_2=0.999, amsgrad=False)

##Adam optimizer.

Default parameters follow those provided in the original paper.

##Arguments

1. learning_rate: float >= 0. Learning rate.
2. beta_1: float, 0 < beta < 1. Generally close to 1.
3. beta_2: float, 0 < beta < 1. Generally close to 1.
4. amsgrad: boolean. Whether to apply the AMSGrad variant of this algorithm from the paper "On the Convergence of Adam and Beyond".

In [0]:
from keras import optimizers
adam=keras.optimizers.Adam(lr=0.001, beta_1=0.9, beta_2=0.999, amsgrad=False)

In [0]:
# Compile model using above optimizer
model.compile(optimizer=adam,
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

#Adamax

keras.optimizers.Adamax(learning_rate=0.002, beta_1=0.9, beta_2=0.999)

Adamax optimizer from Adam paper's Section 7.

It is a variant of Adam based on the infinity norm. Default parameters follow those provided in the paper.

##Arguments

1. learning_rate: float >= 0. Learning rate.
2. beta_1: float, 0 < beta < 1. Generally close to 1.
3. beta_2: float, 0 < beta < 1. Generally close to 1.

In [0]:
from keras import optimizers
adamax=keras.optimizers.Adamax(lr=0.002, beta_1=0.9, beta_2=0.999)


In [0]:
# Compile model using above optimizer
model.compile(optimizer=adamax,
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

#Nadam

keras.optimizers.Nadam(learning_rate=0.002, beta_1=0.9, beta_2=0.999)

##Nesterov Adam optimizer.

Much like Adam is essentially RMSprop with momentum, Nadam is RMSprop with Nesterov momentum.

Default parameters follow those provided in the paper. It is recommended to leave the parameters of this optimizer at their default values.

##Arguments

1. learning_rate: float >= 0. Learning rate.
2. beta_1: float, 0 < beta < 1. Generally close to 1.
3. beta_2: float, 0 < beta < 1. Generally close to 1.

In [0]:
from keras import optimizers
nadam=keras.optimizers.Nadam(lr=0.002, beta_1=0.9, beta_2=0.999)



In [0]:
# Compile model using above optimizer
model.compile(optimizer=nadam,
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

### **Fashion MNIST**

In [0]:
#import packages
import keras
from keras.datasets import fashion_mnist
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation, Flatten
from keras.utils import to_categorical
from keras import optimizers

adamax=keras.optimizers.Adamax(lr=0.002, beta_1=0.9, beta_2=0.999)

#import dataset
(X_train, y_train), (X_test, y_test) = fashion_mnist.load_data()

# Reshape data
X_train = X_train.reshape((X_train.shape[0], 28 * 28))
X_train = X_train.astype('float32') / 255
X_test = X_test.reshape((X_test.shape[0], 28 * 28))
X_test = X_test.astype('float32') / 255

# Categorically encode labels
y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

# Compile model using above optimizer
model.compile(optimizer=adamax,
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=128,
          epochs=10,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

4423680/4422102 [==============================] - 0s 0us/step
Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 34us/step - loss: 0.9270 - acc: 0.7539 - val_loss: 0.5387 - val_acc: 0.8151
Epoch 2/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.4500 - acc: 0.8448 - val_loss: 0.4565 - val_acc: 0.8425
Epoch 3/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.3874 - acc: 0.8643 - val_loss: 0.4110 - val_acc: 0.8536
Epoch 4/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.3524 - acc: 0.8740 - val_loss: 0.3906 - val_acc: 0.8618
Epoch 5/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.3263 - acc: 0.8820 - val_loss: 0.3745 - val_acc: 0.8667
Epoch 6/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.3062 - acc: 0.8896 - val_loss: 0.3671 - val_acc: 0.8704
Epoch 7/10
60000/60000 [==============================] - 2s 27us

### **CIFAR 10**

In [0]:
#import packages
from keras.datasets import cifar10
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation, Flatten
from keras.utils import to_categorical
from keras import optimizers

adamax=keras.optimizers.Adamax(lr=0.002, beta_1=0.9, beta_2=0.999)

#import dataset
(X_train, y_train), (X_test, y_test) = cifar10.load_data()

#change shape from image to vector
X_train = X_train.reshape(50000, 32 * 32 * 3)
X_test = X_test.reshape(10000, 32 * 32 * 3)

#preprocess
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255.0
X_test /= 255.0

#change labels from numeric to one hot encoded
Y_train = to_categorical(y_train, 10)
Y_test =  to_categorical(y_test, 10)

#Model building
model = Sequential()
model.add(Dense(1024, input_shape=(3072, )))
model.add(Activation('relu'))
model.add(Dense(512))
model.add(Activation('relu'))
model.add(Dense(512))
model.add(Activation('relu'))
model.add(Dense(10))
model.add(Activation('softmax'))

# Compile model using above optimizer
model.compile(optimizer=adamax,
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, Y_train,
          batch_size=128,
          epochs=10,
          verbose=1,
          validation_data=(X_test, Y_test))

score = model.evaluate(X_test, Y_test, verbose=1)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 50000 samples, validate on 10000 samples
Epoch 1/10
50000/50000 [==============================] - 3s 54us/step - loss: 2.0352 - acc: 0.2951 - val_loss: 1.7365 - val_acc: 0.3784
Epoch 2/10
50000/50000 [==============================] - 2s 41us/step - loss: 1.7069 - acc: 0.3889 - val_loss: 1.6803 - val_acc: 0.3959
Epoch 3/10
50000/50000 [==============================] - 2s 41us/step - loss: 1.6135 - acc: 0.4257 - val_loss: 1.6326 - val_acc: 0.4214
Epoch 4/10
50000/50000 [==============================] - 2s 43us/step - loss: 1.5376 - acc: 0.4510 - val_loss: 1.4953 - val_acc: 0.4605
Epoch 5/10
50000/50000 [==============================] - 2s 41us/step - loss: 1.4888 - acc: 0.4685 - val_loss: 1.4703 - val_acc: 0.4782
Epoch 6/10
50000/50000 [==============================] - 2s 41us/step - loss: 1.4486 - acc: 0.4831 - val_loss: 1.4599 - val_acc: 0.4809
Epoch 7/10
50000/50000 [==============================] - 2s 41us/step - loss: 1.4065 - acc: 0.4967 - val_loss: 1.4237 - val_acc

### **CIFAR 100**

In [0]:
#import packages
from keras.datasets import cifar100
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation, Flatten
from keras.utils import to_categorical
from keras import optimizers

adamax=keras.optimizers.Adamax(lr=0.002, beta_1=0.9, beta_2=0.999)

#import dataset
(X_train, y_train), (X_test, y_test) = cifar100.load_data()

#change shape from image to vector
X_train = X_train.reshape(50000, 32 * 32 * 3)
X_test = X_test.reshape(10000, 32 * 32 * 3)

#preprocess
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255.0
X_test /= 255.0

#change labels from numeric to one hot encoded
Y_train = to_categorical(y_train, 100)
Y_test =  to_categorical(y_test, 100)

#Model building
model = Sequential()
model.add(Dense(1024, input_shape=(3072, )))
model.add(Activation('relu'))
model.add(Dense(512))
model.add(Activation('relu'))
model.add(Dense(512))
model.add(Activation('relu'))
model.add(Dense(100))
model.add(Activation('softmax'))

# Compile model using above optimizer
model.compile(optimizer=adamax,
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, Y_train,
          batch_size=128,
          epochs=10,
          verbose=1,
          validation_data=(X_test, Y_test))

score = model.evaluate(X_test, Y_test, verbose=1)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

169009152/169001437 [==============================] - 6s 0us/step
Train on 50000 samples, validate on 10000 samples
Epoch 1/10
50000/50000 [==============================] - 3s 53us/step - loss: 4.1948 - acc: 0.0646 - val_loss: 3.8947 - val_acc: 0.1032
Epoch 2/10
50000/50000 [==============================] - 2s 43us/step - loss: 3.7389 - acc: 0.1294 - val_loss: 3.6473 - val_acc: 0.1463
Epoch 3/10
50000/50000 [==============================] - 2s 42us/step - loss: 3.5412 - acc: 0.1624 - val_loss: 3.5357 - val_acc: 0.1740
Epoch 4/10
50000/50000 [==============================] - 2s 42us/step - loss: 3.4031 - acc: 0.1878 - val_loss: 3.4175 - val_acc: 0.1976
Epoch 5/10
50000/50000 [==============================] - 2s 43us/step - loss: 3.2951 - acc: 0.2060 - val_loss: 3.3781 - val_acc: 0.2003
Epoch 6/10
50000/50000 [==============================] - 2s 41us/step - loss: 3.1897 - acc: 0.2246 - val_loss: 3.3264 - val_acc: 0.2102
Epoch 7/10
50000/50000 [==============================] - 2s 

### **IRIS**

In [0]:
#import packages
import pandas
from keras.models import Sequential
from keras.layers import Dense
from keras.wrappers.scikit_learn import KerasClassifier
from keras.utils import np_utils
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import KFold
from sklearn.preprocessing import LabelEncoder
from sklearn.pipeline import Pipeline
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder
from keras import optimizers

adamax=keras.optimizers.Adamax(lr=0.002, beta_1=0.9, beta_2=0.999)

iris_data = load_iris()
	
x = iris_data.data
y_ = iris_data.target.reshape(-1, 1) # Convert data to a single column

# One Hot encode the class labels
encoder = OneHotEncoder(sparse=False)
y = encoder.fit_transform(y_)

# Split the data for training and testing
train_x, test_x, train_y, test_y = train_test_split(x, y, test_size=0.01)

model = Sequential()

model.add(Dense(10, input_shape=(4,), activation='relu'))
model.add(Dense(10, activation='relu'))
model.add(Dense(3, activation='softmax'))

# Compile model using above optimizer
model.compile(optimizer=adamax,
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(train_x, train_y,
          batch_size=5,
          epochs=10,
          verbose=2,
          validation_data=(test_x, test_y))

score = model.evaluate(test_x, test_y, verbose=2)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 148 samples, validate on 2 samples
Epoch 1/10
 - 1s - loss: 2.4086 - acc: 0.2432 - val_loss: 3.8123 - val_acc: 0.0000e+00
Epoch 2/10
 - 0s - loss: 1.5781 - acc: 0.3311 - val_loss: 2.3074 - val_acc: 0.0000e+00
Epoch 3/10
 - 0s - loss: 1.1287 - acc: 0.3378 - val_loss: 1.3651 - val_acc: 0.0000e+00
Epoch 4/10
 - 0s - loss: 0.8598 - acc: 0.5068 - val_loss: 0.8904 - val_acc: 1.0000
Epoch 5/10
 - 0s - loss: 0.7131 - acc: 0.8108 - val_loss: 0.6726 - val_acc: 1.0000
Epoch 6/10
 - 0s - loss: 0.6317 - acc: 0.9595 - val_loss: 0.5321 - val_acc: 1.0000
Epoch 7/10
 - 0s - loss: 0.5817 - acc: 0.9189 - val_loss: 0.4455 - val_acc: 1.0000
Epoch 8/10
 - 0s - loss: 0.5475 - acc: 0.9189 - val_loss: 0.3843 - val_acc: 1.0000
Epoch 9/10
 - 0s - loss: 0.5182 - acc: 0.9662 - val_loss: 0.3340 - val_acc: 1.0000
Epoch 10/10
 - 0s - loss: 0.4935 - acc: 0.9392 - val_loss: 0.3025 - val_acc: 1.0000
Test loss: 0.30251479148864746
Test accuracy: 1.0


Optimizer:


1.   Standard:
acc: 0.9969, 
val_loss: 0.1073, 
val_acc: 0.9794

2.   SGD:
acc: 0.9999, 
val_loss: 0.0829, 
val_acc: 0.9840

3. Rmsprop:
acc: 0.9989,
val_loss: 0.1504, 
val_acc: 0.9814

4. Adagrad:
acc: 1.0000, 
val_loss: 0.1044, 
val_acc: 0.9858

5. Adadelta:
acc: 1.0000, 
val_loss: 0.1058, 
val_acc: 0.9858

6. Adam:
acc: 0.9986, 
val_loss: 0.1008, 
val_acc: 0.9814

7. Adamax:
acc: 1.0000, 
val_loss: 0.0977, 
val_acc: 0.9859

8. Nadam:
acc: 0.9964, 
val_loss: 0.1314, 
val_acc: 0.9796

Conclusion:
After testing all the optimizers, Adamax was found to give the highest validation accuracy.

Fashion MNIST Dataset:
Adamax:
acc: 0.9068, 
val_loss: 0.3395,
val_acc: 0.8813,

CIFAR 10 Dataset:
Adamax:
acc: 0.5355,
val_loss: 1.4160,
val_acc: 0.4966

CIFAR 100 Dataset:
Adamax:
acc: 0.2850,
val_loss: 3.2244,
val_acc: 0.2348

IRIS Dataset:
Adamax: 
acc: 0.9392,
val_loss: 0.3025,
val_acc: 1.0


#Thank you for completing the notebook